sdcm/utils/raft/__init__.py

```diff
@@ -0,0 +1,205 @@
+import contextlib
+import logging
+
+from typing import Protocol
+
+from sdcm.utils.decorators import retrying
+from sdcm.sct_events.database import DatabaseLogEvent
+from sdcm.sct_events.filters import EventsSeverityChangerFilter
+from sdcm.sct_events import Severity
+
+
+LOGGER = logging.getLogger(__name__)
+
+
+class Group0MembersNotConsistentWithTokenRingMembersException(Exception):
+    pass
+
+
+class RaftFeatureOperations(Protocol):
+    _node: 'BaseNode'
+    enabled: bool = False
+
+    def get_status(self):
+        ...
+
+    def is_ready(self):
+        ...
+
+    def get_group0_members(self):
+        ...
+
+    def get_group0_non_voters(self):
+        ...
+
+    def clean_group0_garbage(self, raise_exception: bool = False):
+        ...
+
+    def diff_token_ring_group0_members(self) -> list[str]:
+        ...
+
+
+class Raft(RaftFeatureOperations):
+    def __init__(self, node: "BaseNode") -> None:
+        super().__init__()
+        self._node = node
+        self.enabled = True
+
+    def get_status(self) -> str:
+        """ get raft state """
+        with self._node.parent_cluster.cql_connection_patient_exclusive(node=self._node) as session:
+            row = session.execute("SELECT * FROM system.scylla_local WHERE key = 'group0_upgrade_state'").one()
+            return getattr(row, 'value', "")
+
+    def is_ready(self) -> bool:
+        """check if raft running
+
+        According to https://docs.scylladb.com/branch-5.2/architecture/raft.html
+        """
+        state = self.get_status()
+        return state == "use_post_raft_procedures"
+
+    @retrying(n=5, sleep_time=5, raise_on_exceeded=False)
```

Comment:
What's the point of retrying if this never gonna raise exception ?

Reply:
removed.